engine/src/flutter/lib/web_ui/lib/src/engine/semantics/menus.dart

```diff
@@ -0,0 +1,151 @@
+      ) {
+    setAriaRole('menu');
+  }
+
+  @override
+  bool focusAsRouteDefault() => focusable?.focusAsRouteDefault() ?? false;
+}
+
+/// Indicates a menu bar element.
+///
+/// Uses aria menubar role to convey this semantic information to the element.
+///
+/// Screen-readers takes advantage of "aria-label" to describe the visual.
+class SemanticMenuBar extends SemanticRole {
+  SemanticMenuBar(SemanticsObject semanticsObject)
+    : super.withBasics(
+        EngineSemanticsRole.menuBar,
+        semanticsObject,
+        preferredLabelRepresentation: LabelRepresentation.ariaLabel,
+      ) {
+    setAriaRole('menubar');
+  }
+
+  @override
+  bool focusAsRouteDefault() => focusable?.focusAsRouteDefault() ?? false;
+}
+
+/// Indicates a menu item element.
+///
+/// Uses aria menuitem role to convey this semantic information to the element.
+///
+/// Screen-readers takes advantage of "aria-label" to describe the visual.
+class SemanticMenuItem extends SemanticRole {
+  SemanticMenuItem(SemanticsObject semanticsObject)
+    : super.withBasics(
+        EngineSemanticsRole.menuItem,
+        semanticsObject,
+        preferredLabelRepresentation: LabelRepresentation.ariaLabel,
+      ) {
+    setAriaRole('menuitem');
+    if (semanticsObject.hasExpandedState) {
+      setAttribute('aria-haspopup', 'menu');
+    }
+    if (semanticsObject.hasEnabledState && !semanticsObject.isEnabled) {
+      setAttribute('aria-disabled', 'true');
+    } else {
+      addTappable();
+    }
+  }
+
+  @override
+  void update() {
+    super.update();
+    if (semanticsObject.isFlagsDirty) {
+      if (semanticsObject.hasEnabledState && !semanticsObject.isEnabled) {
+        setAttribute('aria-disabled', 'true');
+      } else {
+        removeAttribute('aria-disabled');
+      }
+    }
+  }
+
+  @override
+  bool focusAsRouteDefault() => focusable?.focusAsRouteDefault() ?? false;
+}
+
+/// Indicates a menu item element with a checkbox.
+///
+/// Uses aria menuitemcheckbox role to convey this semantic information to the element.
+///
+/// Screen-readers takes advantage of "aria-label" to describe the visual.
+class SemanticMenuItemCheckbox extends SemanticRole {
+  SemanticMenuItemCheckbox(SemanticsObject semanticsObject)
+    : super.withBasics(
+        EngineSemanticsRole.menuItemCheckbox,
+        semanticsObject,
+        preferredLabelRepresentation: LabelRepresentation.ariaLabel,
+      ) {
+    setAriaRole('menuitemcheckbox');
+  }
+
+  @override
+  void update() {
+    super.update();
+
+    if (semanticsObject.isFlagsDirty) {
+      String checkedValue;
+      if (semanticsObject.isCheckable && semanticsObject.hasFlag(ui.SemanticsFlag.isChecked)) {
+        checkedValue = 'true';
+      } else if (semanticsObject.isCheckable &&
+          semanticsObject.hasFlag(ui.SemanticsFlag.isCheckStateMixed)) {
+        checkedValue = 'mixed';
+      } else {
+        checkedValue = 'false';
+      }
+      setAttribute('aria-checked', checkedValue);
```

Comment:
does this class also needs disable/enabled flag?

Reply:
we should probably refactor all the dsiabled and enabled code into a semanticsBehavior as well

Reply:
Create a `Disable` SemanticBehavior. Please let me know if there're any naming suggestions:)

Reply:
nit, maybe call `CanDisable`

Reply:
Done!